python/cudf/cudf/core/frame.py

```diff
@@ -319,6 +319,10 @@ def reassign_categories(categories, cols, col_idxs):
         reassign_categories(
             categories, out._index._data, indices[:first_data_column_position]
         )
+        if not isinstance(
+            out._index, cudf.MultiIndex
+        ) and is_categorical_dtype(out._index._values.dtype):
+            out = out.set_index(as_index(out.index._values))
```

Comment:
Can we generalize this to avoid needing to check for a specific `cudf.MultiIndex` object? Also, why do we need to check all the way down to `_index.._values.dtype` instead of just `._index.dtype`?

Reply:
```suggestion
        if not isinstance(
            out._index, cudf.MultiIndex
        ) and isinstance(other._index, cudf.core.index.CategoricalIndex):
            out = out.set_index(as_index(out.index._values))
```
This can be an even light-weight instance check too.

Reply:
This is the `copy_categories` function which is to recreate Categorical objects once libcudf returned integer columns. I wasn't quite sure if the index type would be `CategoricalIndex` here or if it's converted in the `as_index` call below.

Reply:
Got it, `as_index` seems to be handling this all the time when the dtype is a category it redirects to `as_column` which returns `CategoricalColumn` where if the input is `CategoricalColumn` for `as_index` it always builds a `CategoricalIndex`.

Reply:
The problem is that at this point in the code, this isn't actually a `CategoricalIndex`. It comes out of the libcudf call as an `Int8Index`. At that point we want to reassign the categories, but with the way that `reassign_categories` works, it just acts on the underlying `_data` dict under the data/index columns and casts them in place. Basically this is what is happening:

```
>>> x = cudf.core.index.Int8Index([1,2,3])
>>> cats = cudf.Series(['a','b','c'], dtype='category')
>>> x._data[None] = cats._column
>>> x
CategoricalIndex(['a', 'b', 'c'], categories=['a', 'b', 'c'], ordered=False, dtype='category')
>>> type(x)
<class 'cudf.core.index.Int8Index'>
>>> 
```
I didn't see a good way of fixing `reassign_categories` to deal with this issue without removing it, since it doesn't have access to the parent object that it's adjusting the column under it seemed like the longer way around.